Remove redundant calls to constraint functions in matching

wyegelwel opened the pull request:
This is two changes:

1. Cache the list of constraints for a job since getHardConstraints is
called once for each host with available resources
2. Cache the call to needs-gpus? because it is called once for each job
running on each host (for non-gpu hosts)

shamsimam commented on scheduler/src/cook/mesos/constraints.clj:
```diff
@@ -82,7 +79,15 @@
 
-(defrecord gpu-host-constraint [job]
+(defn build-novel-host-constraint
+  "Constructs a novel-host-constraint"
```
Should we also add that the constraint is meant to exclude any hosts the job previously ran on?

shamsimam commented on scheduler/src/cook/mesos/constraints.clj:
```diff
@@ -82,7 +79,15 @@
+                            (mapv :instance/hostname))]
+    (->novel-host-constraint job previous-hosts)))
+
+(defrecord gpu-host-constraint [job needs-gpus?]
```
Would we ever add a `gpu-host-constraint` for a job that does not need gpus?

wyegelwel replied:
Yes, just updated the help message on build-gpu-host-constraint. It is meant to prevent gpu jobs from going on non-gpu hosts and non-gpu jobs from going on gpu hosts

shamsimam commented on scheduler/src/cook/mesos/constraints.clj:
```diff
@@ -93,17 +98,22 @@
                                    "Job does not need gpus, host has gpus."))])))
 
-(def job-constraint-constructors [->novel-host-constraint ->gpu-host-constraint])
+(defn build-gpu-host-constraint
```
+1 for factory methods.

dposada commented on scheduler/src/cook/mesos/constraints.clj:
```diff
@@ -82,7 +79,16 @@
-(defrecord gpu-host-constraint [job]
+(defn build-novel-host-constraint
+  "Constructs a novel-host-constraint. 
+  The constraint prevent the job from running on hosts it has already run on"
```
prevent -> prevent**s**

dposada commented on scheduler/src/cook/mesos/constraints.clj:
```diff
@@ -93,17 +99,24 @@
           vm-has-gpus? (or (get vm-attributes "COOK_GPU?") ; Set when putting attributes in cache
                            (some (fn gpu-task? [req]
-                                   (job-needs-gpus? (:job req)))
+                                   (:needs-gpus? req))
```
Where is `:needs-gpus?` being set on the `req`?

wyegelwel replied:
The req is a TaskRequest, coming from the TaskRequestAdaptor. It is the needs-gpus? argument on the record.

shamsimam replied:
Can we annotate the `req` with its type?

sradack replied:
Also maybe `{:keys [needs-gpus?]}`

dposada commented on scheduler/src/cook/mesos/scheduler.clj:
```diff
@@ -392,14 +392,11 @@
             (str "The machine " (.getHostname target-vm) (if @needs-gpus? " doesn't have" " has") " gpus")))))))
 
-(defrecord TaskRequestAdapter [job resources task-id assigned-resources guuid->considerable-cotask-ids]
+(defrecord TaskRequestAdapter [job resources task-id assigned-resources guuid->considerable-cotask-ids constraints needs-gpus?]
```
Is `needs-gpus?` used anywhere?

wyegelwel replied:
Yes, see the comment above

sradack commented:
LGTM

shamsimam commented:
Needs rebasing.

wyegelwel commented:
rebased. I added two more small changes. The first caches another field (scalar-requests) the second adds type hints to a function that is called for every task,host,constraint so optimizing it is important
